x-pack/plugins/task_manager/server/monitoring/task_run_calcultors.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import stats from 'stats-lite';
+import { JsonObject } from 'src/plugins/kibana_utils/common';
+import { isUndefined, countBy, mapValues } from 'lodash';
+
+export interface AveragedStat extends JsonObject {
+  p50: number;
+  p90: number;
+  p95: number;
+  p99: number;
+}
+
+export function calculateRunningAverage(values: number[]): AveragedStat {
+  return {
+    p50: stats.percentile(values, 0.5),
+    p90: stats.percentile(values, 0.9),
+    p95: stats.percentile(values, 0.95),
+    p99: stats.percentile(values, 0.99),
+  };
+}
+
+/**
+ * Calculate the frequency of each term in a list of terms.
+ * @param values
+ */
+export function calculateFrequency<T>(values: T[]): JsonObject {
+  return values.length
+    ? mapValues(countBy(values), (count) => Math.round((count * 100) / values.length))
+    : {};
+}
+
+/**
+ * Utility to keep track of a limited queue of values which changes over time
+ * dropping older values as they slide out of the window we wish to track
+ */
+export function createRunningAveragedStat<T>(runningAverageWindowSize: number) {
+  const queue = new Array<T>();
+  return (value?: T) => {
+    if (isUndefined(value)) {
+      return queue;
+    } else {
+      if (queue.length === runningAverageWindowSize) {
+        queue.shift();
+      }
+      queue.push(value);
+      return [...queue];
```

Comment:
No, they are actually different - the first is returning the current array, the second is meant to return a clone that won't be mutate, as  *we* mutate it every time we pus ha value in.
The idea is to return a snapshot of what it was at the time, otherwise new values will mutate the returned array.

Looking at this again I feel this could be clearer - I'll look into refactoring this to make it easier to maintain.

Reply:
The thinking in returning a non cloned array in the first return was to avoid creating *another* array in memory just because I happened to know where that's used, but longer term, that could cause bugs if another developer doesn't notice the danger, so I'll clone here and we can optimise this in the future if we encounter memory issues.